Copy file to owner's trash when recipient moves out of share

PVince81 opened the pull request:
## Description
Whenever a share recipient moves files or folders out of the share,
make a backup copy in the owner's trash just in case.

The metadata stays on the recipient's copy that was moved out.

## Related Issue
Fixes https://github.com/owncloud/core/issues/24053 (without activity entry)

## Motivation and Context
Solves the case where the recipient is not reachable any more so the owner still has a chance to restore the file.

## How Has This Been Tested?

1. Login as "user0" (owner)
1. Create a folder "shared"
1. Share "shared" with "user1" (recipient)
1. Login as "user1"
1. Create a file "shared/test.txt" and overwrite several times to create versions
1. Create a folder "shared/sub"
1. Create a file "shared/sub/testsub.txt" and overwrite several times to create versions
1. Rename "shared" to "shared_renamed" (this covers the path matching logic)
1. Move "sub" and "test.txt" out of the "shared_renamed" folder.
1. Login as "owner"
1. Check that trashbin contains both "sub/testsub.txt" and "test.txt"
1. Restore both
1. Check that both "test.txt" and "sub/testsub.txt" are restored in the "shared" folder (from the owner's perspective)
1. Check that both entries still have their versions attached

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## TODOs:

- [x] find a way to bypass the locking issue or find another place / hook to do this operation...

Commit:
Copy file to owner's trash when recipient moves out of share

Whenever a share recipient moves files or folders out of the share,
make a backup copy in the owner's trash just in case.

The metadata stays on the recipient's copy that was moved out.

PVince81 commented:
Currently when moving out "test/deletedk.txt" into the root as recipient:
```
{"reqId":"gpWkNw+rulBC4FrPhDt5","remoteAddr":"127.0.0.1","app":"webdav","message":"Exception: {\"Message\":\"HTTP\\\/1.1 423 \\\"test\\\/deletedk.txt\\\" is locked\",\"Exception\":\"OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Exception\\\\FileLocked\",\"Code\":0,\"Trace\":\"#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(642): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\ObjectTree->move('test\\\/deletedk.t...', 'deletedk.txt')\\n#1 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpMove(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#2 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(479): Sabre\\\\Event\\\\EventEmitter->emit('method:MOVE', Array)\\n#4 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v1\\\/webdav.php(58): Sabre\\\\DAV\\\\Server->exec()\\n#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/remote.php(164): require_once('\\\/srv\\\/www\\\/htdocs...')\\n#7 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/ObjectTree.php\",\"Line\":271,\"User\":\"user1\"}","level":4,"time":"2017-01-27T11:38:59+00:00","method":"MOVE","url":"\/owncloud\/remote.php\/webdav\/test\/deletedk.txt","user":"user1"}
```

PVince81 commented:
It looks like I won't be able to bypass the locking issue from this point of the code. The problem is that the `View` class already sets an exclusive lock earlier and it wouldn't be suitable to change that: https://github.com/owncloud/core/blob/v9.1.3/lib/private/Files/View.php#L767

So the only solution is to do the copy in question from within the pre rename hook, in which only a shared lock is applied.

PVince81 commented:
- [ ] TODO: find out who has the versions in a "recipient deletes" scenario

PVince81 commented:
- [ ] REQUIRES: https://github.com/owncloud/core/pull/27069

PVince81 commented:
- [ ] make sure fed sharing doesn't break in this scenario due to "getOwner"

PVince81 commented:
- [ ] fix versions case
- [ ] add unit tests + verify if versions were stored as well
- [ ] integration test with versions ? (bootstrap missing...)

PVince81 commented:
- [ ] test where the shared folder name is different, in case of potential bugs

PVince81 commented:
- [ ] TODO: documentation ticket for 10.0

PVince81 commented:
It's looking good and also works already.

- The integration test covers only detecting if the file is in trashbin.
- Some existing tests from https://github.com/owncloud/core/pull/27069 ensure that the favorite info is kept, which confirms file id preservation for the recipient. The backup gets a new fileid but no metadata except versions.
- Versions are copied to owner's trash backup just like for delete. This is not covered by integration tests (we lack version code there...) but is covered by the unit tests

PVince81 commented:
This is now ready for review @jvillafanez @VicDeo @DeepDiver1975 @IljaN

PVince81 commented:
Tests pass for me locally 😢

PVince81 commented:
Rebasing as a tentative of luck

PVince81 commented:
Failed again. The tests do seem related.

But let's make sure we got a clean env with a separate Jenkins PR: https://github.com/owncloud/core/pull/27079

PVince81 commented:
Same failure on a separate branch.

Will need to setup an env that is closer to Jenkins to reproduce this.

PVince81 commented:
From Jenkins:
```
17:11:42 There were 11 failures:
17:11:42 
17:11:42 1) TrashbinTest::testExpireOldFiles
17:11:42 Failed asserting that 0 is identical to 3.
17:11:42 
17:11:42 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/apps/files_trashbin/tests/TrashbinTest.php:170
17:11:42 
17:11:42 2) TrashbinTest::testExpireOldFilesShared
17:11:42 Failed asserting that 0 is identical to 3.
17:11:42 
17:11:42 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/apps/files_trashbin/tests/TrashbinTest.php:229
17:11:42 
17:11:42 3) TrashbinTest::testExpireOldFilesUtilLimitsAreMet
17:11:42 Failed asserting that 0 is identical to 3.
17:11:42 
17:11:42 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/apps/files_trashbin/tests/TrashbinTest.php:334
17:11:42 
17:11:42 4) TrashbinTest::testRestoreFileInRoot
17:11:42 Failed asserting that actual size 0 matches expected size 1.
17:11:42 
17:11:42 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/apps/files_trashbin/tests/TrashbinTest.php:363
17:11:42 
17:11:42 5) TrashbinTest::testRestoreFileInSubfolder
17:11:42 Failed asserting that actual size 0 matches expected size 1.
17:11:42 
17:11:42 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/apps/files_trashbin/tests/TrashbinTest.php:396
17:11:42 
17:11:42 6) TrashbinTest::testRestoreFolder
17:11:42 Failed asserting that actual size 0 matches expected size 1.
17:11:42 
17:11:42 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/apps/files_trashbin/tests/TrashbinTest.php:429
17:11:42 
17:11:42 7) TrashbinTest::testRestoreFileFromTrashedSubfolder
17:11:42 Failed asserting that actual size 0 matches expected size 1.
17:11:42 
17:11:42 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/apps/files_trashbin/tests/TrashbinTest.php:462
17:11:42 
17:11:42 8) TrashbinTest::testRestoreFileWithMissingSourceFolder
17:11:42 Failed asserting that actual size 0 matches expected size 1.
17:11:42 
17:11:42 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/apps/files_trashbin/tests/TrashbinTest.php:496
17:11:42 
17:11:42 9) TrashbinTest::testRestoreFileDoesNotOverwriteExistingInRoot
17:11:42 Failed asserting that actual size 0 matches expected size 1.
17:11:42 
17:11:42 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/apps/files_trashbin/tests/TrashbinTest.php:532
17:11:42 
17:11:42 10) TrashbinTest::testRestoreFileDoesNotOverwriteExistingInSubfolder
17:11:42 Failed asserting that actual size 0 matches expected size 1.
17:11:42 
17:11:42 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/apps/files_trashbin/tests/TrashbinTest.php:573
17:11:42 
17:11:42 11) TrashbinTest::testRestoreFileIntoReadOnlySourceFolder
17:11:42 Failed asserting that actual size 0 matches expected size 1.
17:11:42 
17:11:42 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/apps/files_trashbin/tests/TrashbinTest.php:627
```

PVince81 commented:
Aha:
```
PHP 7.1.0RC4
```

might be a clue...

PVince81 commented:
Ran all tests with PHP 7.1.0RC4 but there was no such failure.

@DeepDiver1975 by the way, we should update Jenkins to use the released version 7.1.0

PVince81 commented:
- [ ] find out if the failing test is a side effect of the added test
- [ ] find out if the failing test is caused by the storage changes

PVince81 commented:
Ok, so disabling the newly added unit test makes all tests pass.

So let's hunt for potential side effects...

PVince81 commented:
Another try: properly delete the recipient user in the new test.

So far the side effect seems that trashbin is disabled when running TrashbinTest because any deletion doesn't seem to put any files to trash.

PVince81 commented:
Ok, let's add code to explicitly enable trashbin in the trashbin test...

PVince81 commented:
let's try removing 7.1 to see if it's specific to that PHP version...

PVince81 commented:
It's not 7.1.

Let's now remove pieces of the test that causes side effects... hopefully to find which piece causes it

PVince81 commented:
Failure so far. Let's comment out even more code.

PVince81 commented:
Aha, did not fail...

PVince81 commented:
Still trying to comment out blocks to find out what is causing the failure on Jenkins.

PVince81 commented:
Ok... so not moving files out of the share doesn't cause the side effect. That's a little clue.

PVince81 commented:
Failed again 😢

PVince81 commented:
The trashbin is indeed enabled in TrashbinTest and yet after deleting files, they do not appear in the trashbin at all. WTF

PVince81 commented:
Ah no, wait:
```
TrashbinTest::testExpireOldFiles
Failed asserting that true is false.

/var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/apps/files_trashbin/tests/TrashbinTest.php:161
```

`$disableTrash` is still true when reaching that test...

Investigating why it's not set back...

PVince81 commented:
It turns out that the rename that moves out the file out of the share fails... and because it fails it doesn't call the post rename hook which is supposed to reenable trashbin globally...

Now why does the rename of "/recipient_GDmZ3Qg44VRY9/files/share_renamed/test.txt" to "/recipient_GDmZ3Qg44VRY9/files/test.txt" fail...

```
$result = $storage2->moveFromStorage($storage1, $internalPath1, $internalPath2);
```
with
```
// by property_get
$storage1                        = (object|OCA\Files_Trashbin\Storage[10]);
  $storage1::disableTrash        = (bool) '1';
  $storage1->mountPoint          = (string[37]) '/recipient_1SaLv97GJHfsL/files/share/';
  $storage1->deletedFiles        = (array[0]);
  $storage1->userManager         = (object|OC\User\Manager[4])+;
  $storage1->storage             = (object|OCA\Files_Sharing\SharedStorage[15])+;
  $storage1->cache               = (null);
  $storage1->scanner             = (null);
  $storage1->watcher             = (null);
  $storage1->propagator          = (null);
  $storage1->updater             = (null);


// by property_get
$storage2                        = (object|OCA\Files_Trashbin\Storage[10]);
  $storage2::disableTrash        = (bool) '1';
  $storage2->mountPoint          = (string[25]) '/recipient_1SaLv97GJHfsL/';
  $storage2->deletedFiles        = (array[0]);
  $storage2->userManager         = (object|OC\User\Manager[4])+;
  $storage2->storage             = (object|OC\Files\Storage\Home[15])+;
  $storage2->cache               = (null);
  $storage2->scanner             = (object|OC\Files\Cache\Scanner[7])+;
  $storage2->watcher             = (null);
  $storage2->propagator          = (null);
  $storage2->updater             = (object|OC\Files\Cache\Updater[5])+;


// by property_get
$internalPath1                   = (string[8]) 'test.txt';


// by property_get
$internalPath2                   = (string[14]) 'files/test.txt';
```

The paths used for the rename in the Local storage looks really strange:
- source: files/share/test.txt
- target: /dev/shm/data-autotest/recipient_1SaLv97GJHfsL/files/test.txt

instead of
- source: /dev/shm/data-autotest/user_xyz/files/share/test.txt
- target: /dev/shm/data-autotest/recipient_1SaLv97GJHfsL/files/test.txt

PVince81 commented:
So it turns out that:
- when doing this kind of cross storage move in the web UI, it gets translated to copy + unlink. It doesn't detect that they are on the same absolute local storage
- when doing this in the unit test, it detects that it's on the same storage and does a rename internally, which fails due to wrong paths...

Now I wonder if the latter was always broken before and never noticed...

PVince81 commented:
Yeah... when the encryption wrapper is enabled, it always does a copy + unlink for `moveFromStorage`.
But that wrapper is not applied in the unit tests.

PVince81 commented:
Arghhhhh, disabling the encryption wrapper makes cross-storage move fail. It was a hidden bug 😢

PVince81 commented:
Raised https://github.com/owncloud/core/issues/27124

- [ ] BLOCKED by https://github.com/owncloud/core/issues/27124

jvillafanez commented on apps/files_trashbin/lib/Trashbin.php:
```diff
@@ -293,25 +320,30 @@ public static function move2trash($file_path) {
 
 			if ($rootView->is_dir($owner . '/files_versions/' . $ownerPath)) {
-				if ($owner !== $user) {
+				if ($owner !== $user || $forceCopy) {
```
Maybe move these checks to another function? It might remove the duplication here and also simplify this function

PVince81 replied:
I prefer to not touch too much here and just inject what I need.

This would be a good task for future refactoring/rewrite of this app with proper tests.

PVince81 commented:
Let's see... maybe I can just register an encryption wrapper in the tests... at least this will bring all tests closer to reality.

Commit:
Register encryption storage wrapper in trashbin storage test

Because the encryption wrapper does some twisted stuff with methods like
moveFromStorage, we need to apply it here to make sure the behavior is
close to reality.

PVince81 commented:
Here we go, the encryption wrapper fixes the test: https://github.com/owncloud/core/pull/27042/commits/e22d634673da29d61aa94cb1646955057d00bef0

jvillafanez commented:
I don't see any other change that should be done, so as long as it's working 👍

PVince81 commented:
So... now StorageTests pass but TrashbinTest fail again 😢

PVince81 commented:
Failure on pgsql now, so much fun:

```
19:35:56 1) OCA\Files_Trashbin\Tests\StorageTest::testOwnerBackupWhenMovingFileOutOfShare
19:35:56 Doctrine\DBAL\Exception\DriverException: An exception occurred while executing 'SELECT "path", "fileid" FROM "oc_filecache" WHERE "storage" = ? AND "path" LIKE ?' with params [false, "sub\/%"]:
19:35:56 
19:35:56 SQLSTATE[22P02]: Invalid text representation: 7 ERROR:  invalid input syntax for integer: ""
19:35:56 
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/composer/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractPostgreSQLDriver.php:91
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/composer/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php:128
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/composer/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:836
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/private/DB/Connection.php:192
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/private/Files/Cache/Cache.php:514
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/private/Files/Cache/Updater.php:197
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/private/Files/View.php:325
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/private/Files/View.php:802
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/private/Files/Node/Node.php:414
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/apps/files_trashbin/tests/StorageTest.php:496
19:35:56 
19:35:56 Caused by
19:35:56 Doctrine\DBAL\Driver\PDOException: SQLSTATE[22P02]: Invalid text representation: 7 ERROR:  invalid input syntax for integer: ""
19:35:56 
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/composer/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:93
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/composer/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:830
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/private/DB/Connection.php:192
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/private/Files/Cache/Cache.php:514
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/private/Files/Cache/Updater.php:197
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/private/Files/View.php:325
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/private/Files/View.php:802
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/private/Files/Node/Node.php:414
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/apps/files_trashbin/tests/StorageTest.php:496
19:35:56 
19:35:56 Caused by
19:35:56 PDOException: SQLSTATE[22P02]: Invalid text representation: 7 ERROR:  invalid input syntax for integer: ""
19:35:56 
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/composer/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:91
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/composer/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:830
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/private/DB/Connection.php:192
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/private/Files/Cache/Cache.php:514
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/private/Files/Cache/Updater.php:197
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/private/Files/View.php:325
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/private/Files/View.php:802
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/lib/private/Files/Node/Node.php:414
19:35:56 /var/lib/jenkins/workspace/owncloud-core_core_PR-27042-D3EHZDNEIUXO3TN26UHXFNK7YRY4G5SCHC33PRCY6GB2OV74TYEA/apps/files_trashbin/tests/StorageTest.php:496
19:35:56 
```

PVince81 commented:
Looks like the source storage id is `(bool)0` before the rename in the test.
Somehow SQLite and MySQL don't mind about this...

Need to make sure that the source storage exists in DB before doing this.

PVince81 commented:
Arghhh, seems there is yet another bug in the FS code.

The SharedCache doesn't seem to forward all calls to the underlying cache, so it cannot get its numeric id.

PVince81 commented:
Looks like this code was completely wrong. There isn't even a `numericId` attribute neither in this class nor in the parent classes.

But so far no unit test failed because of this...

PVince81 commented:
I made a separate PR for this fix with an extra unit test: https://github.com/owncloud/core/pull/27172

PVince81 commented:
I have now cherry-picked the fix from https://github.com/owncloud/core/pull/27172 to this PR here.

Let's see if the tests pass first and then make a decision whether to keep the extra fix or not.

PVince81 commented:
The extra fix has evolved a bit.

Waiting for https://github.com/owncloud/core/pull/27172 to be merged then rebase.

PVince81 commented:
Rebased to include the final fix from https://github.com/owncloud/core/pull/27172

PVince81 commented:
GREEN ! FINALLY GREEN ! After all these struggles.

I hope you'll enjoy this enhancement, guys...

PVince81 commented:
Added to new features of 10.0: https://github.com/owncloud/core/wiki/ownCloud-10.0-Features

PVince81 commented:
Regression: https://github.com/owncloud/core/issues/27698
